File: py.fcn/fitfcn.py
"""
Module fitfcn
=============
version 1.1
-------------

List of function focused on the parameterisation and curve fitting.
contains:
    - phot
    - residuals
    - fitTUV
    - fitStat
"""

import sys
#______________________________________________________________________________________________

def phot(p,x):
    """
        Function phot
        =============

        Purpose:
            Parameterisation for photoreactions in MCM.
            Parameterise sza dependent j values from TUV model calculations.

        Variables:
        I/O:
            p(l,m,n):   array of parameter l (in s-1), m, n
            x:          sza (independent variable) in parameterisation
            jval:       j values (dependent variable in parameterisation)

        internal:


        Dependencies:
        uses:           numpy
        called from:    photMCM (main)
        """
    l,m,n=p
    import numpy as np
    jval = l*((np.cos(x))**m)*np.exp(-n/np.cos(x))
    return jval
#______________________________________________________________________________________________


def residuals(p,xdata,jval):
    """
        Function residuals
        ==================

        Purpose:
            =

        Variables:
        I/O:
            p(l,m,n):   fit paramters l (in s-1), m, n
            xdata:      sza (values of independent variable)
            jval:       j values (values of dependent variable)

        Dependencies:
        uses:           fitfcn.phot
        called from:    fitfcn.fitStat, scipy.optimize.leastsq in fitfcn.fitTUV
    """

    return jval - phot(p,xdata)
#______________________________________________________________________________________________


def fitTUV(xdata,ydata):
    """
    Function fitTUV
    ===============

    Purpose:
        Curve fit of TUV data to parameterisation for photolysis in MCM.8///////////////////8=74
        j = l*(cos(x))^m*exp(-n*sec(x))

    Variables:
    I/O:
        xdata,ydata:    x-,y-data for curve fit (sza and j values)
        p:              optimised parameters from least square fit
        cov:            covariance matrix from least square fit
        infodict:       further statistical data from least square fit
        mesg, ier:      not used (see description of leastsq function)

    internal:
        Param           list with parameters for curve fitting
        p_guess:        initial parameters for curve fit

    Dependencies:
        uses:           residuals,numpy,scipy.optimize.leastsq, collections.namedtuple
        called from:    datfcn.xydat
    """
    # load functions
    import numpy as np
    import collections #
    from scipy.optimize import leastsq # for curve fit
    from datfcn import order

# curve fitting
    Param=collections.namedtuple('Param','l m n')
    p_guess=Param(l=ydata[0]*np.exp(-0.8),m=0.8,n=0.25)
    p,cov,infodict,mesg,ier = leastsq(residuals,p_guess,args=(xdata,ydata),full_output=True)
    p=Param(*p)
    l = p[0]
    ol, ml = order(l)
    l = l/ml
    m = p[1]
    n = p[2]
    return p,l,ol,m,n,cov,infodict
#______________________________________________________________________________________________


def fitStat(xdata,ydata,p,cov,infodict):
    """
    Function fitStat
    ================

    Purpose:
        Derive statistical data from least square fit of TUV data for MCM parameterisations.

    Variables:
    I/O:
        xdata,ydata:    x-,y-data for curve fit (sza and j values)
        p:              optimised parameters from least square fit
        cov:            covariance matrix from least square fit
        infodict:       further statistical data from least square fit
        rsquared:       correlation coefficient
        ss_tot:         standard deviation
        rmse:           root mean square error
        el,em,en:       confidence of parameters

    internal:
        ss_err:         for calculation of R2 and RMSE
        dof:            degrees of freedom for calculation of RMSE
        s_sq:           for calculation of confidences of parameters
        err:            list of all confidences

    Dependencies:
        uses:           residuals,datfcn.order,numpy
        called from:    datfcn.xydat
        """
    # load functions
    import numpy as np
    from datfcn import order


    # R2:
    ss_err=(infodict['fvec']**2).sum()
    ss_tot=((ydata-ydata.mean())**2).sum()
    rsquared=1-(ss_err/ss_tot)

    # RMSE:
    dof=len(xdata)-len(p)
    rmse=np.sqrt(ss_err/dof)

    # Confidence:

    if (len(ydata) > len(p)) and cov is not None:
        s_sq = (residuals(p, xdata, ydata)**2).sum()/(len(ydata)-len(p))
        cov = cov * s_sq
    else:
        cov = float("inf")
    try:
        err = np.sqrt(np.diag(cov))
        el = err[0] / order(p[0])[1]
        em = err[1]
        en = err[2]
    except:
        err = float("inf")
        el  = float("inf")
        em  = float("inf")
        en  = float("inf")


    return rsquared,ss_tot,rmse,el,em,en
#______________________________________________________________________________________________

File: jl.mod/fitfcn.jl
__precompile__()


"""
# Module fitfcn

Fit MCM parameterisations to TUV output:
    j / s^-1 = l·cos^m(χ)·exp(-n·sec(χ))
"""
module fitfcn

using DataFrames
using LsqFit
export fit_j


"""
    fit_j(jvals)

Derive MCM parameterisations for dataframe jvals with χ-dependent _j_ values.
    j / s^-1 = l·cos^m(χ)·exp(-n·sec(χ))
"""
function fit_j(jvals)

  # Define x values and transform from deg to rad
  xdata = jvals[:sza].data
  xdata = xdata.* π./180.
  # Initialise arrays for fitting data
  fit = []; sigma = []; rmse = []; R2 = []

  delrxn = Int64[]
  # Loop over all j data
  for i = 2:length(jvals)
    ydata = jvals[i].data # define y data
    if ydata[1] == 0.
      println("\033[95mReaction with no data skipped:\033[0m")
      println(convert(String,names(jvals)[i]))
      push!(delrxn,i)
      continue
    end

    # Define fit function with initial guesses
    model(x,p) = p[1].*(cos.(x)).^(p[2]).*exp.(-p[3]./cos.(x))
    p0 = [ydata[1],0.8,0.3]

    # Derive fit
    push!(fit, curve_fit(model, xdata, ydata, p0))
    # Derive sigma with 95% confidence
    push!(sigma, estimate_errors(fit[i-1-length(delrxn)],0.95))
    # Calculate statistical data for RMSE and R^2
    ss_err = sum(fit[i-1-length(delrxn)].resid.^2)
    ss_tot = sum((ydata-mean(ydata)).^2)
    # RMSE
    push!(rmse, √(ss_err/fit[i-1-length(delrxn)].dof))
    # R^2
    push!(R2, 1. - (ss_err/ss_tot))
  end
  for i in delrxn  delete!(jvals, i)  end

  return fit, sigma, rmse, R2

end #function fit_j

end #module fitfcn
